include/quic/connection.hpp

```diff
@@ -13,137 +13,150 @@
 #include <uvw.hpp>
 
 #include "context.hpp"
-#include "crypto.hpp"
+#include "gnutls_crypto.hpp"
 #include "utils.hpp"
 
 namespace oxen::quic
 {
-    class Endpoint;
-    class Stream;
-    class Server;
-    class Client;
-    class Handler;
+    /*
+        TODO:
+        - tls creds and session?
+        - user config settings (for debugging)
+        - closing connection through conn interface?
+            - current close mechanism lives in endpoint
+
+        - tests
+            - stream test cases for switching # of streams mid connection
+            - fail cases        
+    */
+
+    class connection_interface
+    {
+      public:
+        virtual std::shared_ptr<Stream> get_new_stream(
+                stream_data_callback_t data_cb = nullptr, stream_close_callback_t close_cb = nullptr) = 0;
+
+        virtual const ConnectionID& scid() = 0;
+    };
 
-    class Connection
+    class Connection : public connection_interface, public std::enable_shared_from_this<Connection>
     {
+      public:
+        // Construct and initialize a new inbound/outbound connection to/from a remote
+        //      ep: owning endpoints
+        //      scid: local ("primary") CID used for this connection (random for outgoing)
+        //		dcid: remote CID used for this connection
+        //      path: network path used to reach remote client
+        //		handle: udp handle dedicated to local address
+        //		creds: relevant tls information per connection
+        //		u_config: user configuration values passed in struct
+        //      dir: enum specifying configuration detailts for client vs. server
+        //		hdr: optional parameter to pass to ngtcp2 for server specific details
+        Connection(
+                Endpoint& ep,
+                const ConnectionID& scid,
+                const ConnectionID& dcid,
+                const Path& path,
+                std::shared_ptr<uv_udp_t> handle,
+                std::shared_ptr<ContextBase> ctx,
+                Direction dir,
+                ngtcp2_pkt_hd* hdr = nullptr);
+        ~Connection();
+
+        static std::shared_ptr<Connection> make_conn(
+                Endpoint& ep,
+                const ConnectionID& scid,
+                const ConnectionID& dcid,
+                const Path& path,
+                std::shared_ptr<uv_udp_t> handle,
+                std::shared_ptr<ContextBase> ctx,
+                Direction dir,
+                ngtcp2_pkt_hd* hdr = nullptr);
+
+        // Returns a pointer to the owning endpoint, else nullptr
+        Endpoint* endpoint();
+        const Endpoint* endpoint() const;
+
+        void io_ready();
+
+        const GNUTLSSession* get_session()
```

Comment:
Connection has no idea which sort of TLSSession it owns and should not care; that's the purpose of abstracting TLSCreds and TLSSession.

Ok right *now* it knows it's a GNUTLSSession but that's because that's the only one implemented...in general it shouldn't know or care.